sysl2/sysl/ints.go

```diff
@@ -0,0 +1,215 @@
+			v.handleStatement(appname, epname, t.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			v.handleStatement(appname, epname, t.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			v.handleStatement(appname, epname, t.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			v.handleStatement(appname, epname, t.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range t.Alt.GetChoice() {
+				v.handleStatement(appname, epname, choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func (v *IntegrationParam) addAppDependency(sourceApp, sourceEndpt, targetApp, targetEndpt string, stmt *sysl.Statement) bool {
+	if len(v.apps) == 0 {
+		v.apps = v.apps.Union(v.drawableApps)
+	}
+	if sourceEndpt == ".. * <- *" || sourceEndpt == "*" {
+		return false
+	}
+	k := fmt.Sprintf("%s:%s:%s:%s", sourceApp, sourceEndpt, targetApp, targetEndpt)
+	if _, has := v.deps[k]; has {
+		return false
+	}
+	if v.excludes.Contains(sourceApp) || v.excludes.Contains(targetApp) {
+		return false
+	}
+	if v.drawableApps.Contains(sourceApp) || v.drawableApps.Contains(targetApp) {
+		if !hasPattern(v.m.GetApps()[sourceApp].GetAttrs(), "human") {
+			v.apps.Insert(sourceApp)
+		}
+		if !hasPattern(v.m.GetApps()[targetApp].GetAttrs(), "human") {
+			v.apps.Insert(targetApp)
+		}
+	}
+	apps := MakeStrSet(sourceApp, targetApp)
+	if !(apps.IsSubSet(v.apps) || (v.apps.Contains(sourceApp) && v.passthroughs.Contains(targetApp)) || v.passthroughs.Contains(sourceApp)) {
+		return false
+	}
+	v.deps[k] = MakeAppDependency(MakeAppElement(sourceApp, sourceEndpt), MakeAppElement(targetApp, targetEndpt), stmt)
+
+	return true
+}
+
+func GenerateIntegrations(
+	root_model, title, output, project, filter, modules string,
+	exclude []string, clustered, epa bool,
+) map[string]string {
+	r := make(map[string]string)
+	if filter != "" {
+		re := regexp.MustCompile(filter)
+		if !re.MatchString(output) {
+			return r
+		}
+	}
+	mod := loadApp(root_model, modules)
+
+	if len(exclude) == 0 && project != "" {
+		exclude = []string{project}
+	}
+	excludeStrSet := MakeStrSet(exclude...)
+
+	// The "project" app that specifies the required view of the integration
+	app := mod.GetApps()[project]
+	of := MakeFormatParser(output)
+	// Interate over each endpoint within the selected project
+	for epname, endpt := range app.GetEndpoints() {
+		excludes := MakeStrSetFromSpecificAttr("exclude", endpt.GetAttrs())
+		passthroughs := MakeStrSetFromSpecificAttr("passthrough", endpt.GetAttrs())
+		integrations := MakeStrSetFromActionStatement(endpt.GetStmt())
+		output_dir := of.FmtOutput(project, epname, endpt.GetLongName(), endpt.GetAttrs())
+
+		drawableApps := findDrawableApps(mod, integrations)
+		v := MakeIntegrationParam(mod, excludeStrSet.Union(excludes), passthroughs, drawableApps)
+		v.collectApplicationDependencies(integrations)
+		intsParam := &IntsParam{v.apps.ToSlice(), v.drawableApps, v.deps, app, endpt}
+		args := &Args{title, project, clustered, epa}
```

Comment:
`IntegrationParam` and `IntsParam` is confusing.

You could remove IntegrationParam like so:

apps, drawableApps, dependencies := collectApplicationDependencies (excludes, passthrough, integtrations)
This makes the dependencies of function `collectApplicationDependencies` clear.

Reply:
I think the change is not good. Here the better way is to rename the IntegrationParam. If I change the collectApplicationDependencies method, the method inside it would take long parameters. It looks not good.

Reply:
receiving multiple arguments was invented precisely to avoid inventing structures like `IntegrationDependency `.